catalog/IndexScheme.hpp

```diff
@@ -0,0 +1,193 @@
+   *
+   * @return The number of indices defined for the relation.
+   **/
+  inline const std::size_t getNumIndices() const {
+    return index_map_.size();
+  }
+
+  /**
+   * @brief Check whether an index with the given exists or not.
+   *
+   * @param index_name Name of the index to be checked.
+   * @return Whether the index exists or not.
+   **/
+  bool hasIndexWithName(const std::string &index_name) const {
+    return index_map_.find(index_name) != index_map_.end();
+  }
+
+  /**
+   * @brief Check whether an index with the given description
+   *        containing the same attribute id and index type
+   *        exists or not in the index map.
+   *
+   * @param index_descripton Index Description to check against.
+   * @return Whether a similar index description was already defined or not.
+   **/
+  bool hasIndexWithDescription(const IndexSubBlockDescription &index_description) const {
+    // Iterate through every vector of descriptions corresponding to all keys in the index map
+    for (auto cit = index_map_.cbegin(); cit != index_map_.cend(); ++cit) {
+      std::vector<IndexSubBlockDescription> index_descriptions = cit->second;
+      for (auto index_description_itr = index_descriptions.cbegin();
+           index_description_itr != index_descriptions.cend();
+           ++index_description_itr) {
+        const IndexSubBlockDescription &stored_description = *index_description_itr;
+        // check if the stored description matches as the given description
+        if (areIndexDescriptionsSimilar(stored_description, index_description)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  /**
+   * @brief Check whether two index descriptions are similar or not.
+   *        Two index descriptions are similar if they have same
+   *        attributes ids and same index type.
+   *
+   * @param desc_a First index description.
+   * @param desc_b Second index description.
+   * @return Whether the two index_descriptions are similar or not
+   **/
+  bool areIndexDescriptionsSimilar(const IndexSubBlockDescription &desc_a,
+                                   const IndexSubBlockDescription &desc_b) const {
+    if (desc_a.sub_block_type() != desc_b.sub_block_type()) {
+      return false;
+    }
+    // TODO(@ssaurabh): add condition to check for matching attribute ids for both indices
+    return true;
+  }
+
+  /**
+   * @brief Adds a new index entry to the index map.
+   * @warning Must call before hasIndexWithName() and hasIndexWithDescription().
+   *
+   * @param index_name The name of index to add (key)
+   * @param index_description The index description for this index (value)
+   **/
+  bool addIndexMapEntry(const std::string &index_name,
+                        const std::vector<IndexSubBlockDescription> &index_descriptions) {
+    if (index_map_.find(index_name) != index_map_.end()) {
+      return false;  // index_name is already present!
+    } else {
+      std::string key = index_name;
```

Comment:
Remove this line, and use `index_name` directly to avoid a deep copy of strings.